src/types.rs

```diff
@@ -48,11 +96,23 @@ impl TypeRef {
                 BuiltInTypeDefinition::Derived(other_type) => other_type.name().to_owned().unwrap(),
             },
         };
+
+        // convert given IonSchemaElement to an OwnedElement
+        // converts document type to SExpression
+        let value = match value.into() {
+            IonSchemaElement::Element(element) => element,
+            IonSchemaElement::Document(document) => {
+                OwnedElement::new_sexp(document.iter().cloned())
+                    .with_annotations(vec![text_token("document")])
+            }
+        };
+
         for constraint in type_def.constraints() {
-            if let Err(violation) = constraint.validate(value, &self.type_store) {
+            if let Err(violation) = constraint.validate(&value, &self.type_store) {
```

Comment:
issue—as I mentioned in another comment, I think you have an escaping problem if you pass around documents as annotated s-expressions. I believe you can solve it by changing the signature of the validate functions to accept an `IonSchemaElement` instead of an `OwnedElement`.